### DISCUSSION TITLE
How to install home-brew from a python script?

### DISCUSSION THREAD

**Discussion by gauthierbuttez** [Homebrew/discussions#2] [Everyday usage]
**Environment**

Python 3.7

Mac OS Catalina 10.15

**Problem**

This makes me crazy!

I need to install home-brew from my python script:

When I run:

    #mac_account is actual user who is not root
    f'sudo -u {mac_account} /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"'

I get this error output:

    Need sudo access on macOS

When I run the command as Admin:

    /usr/bin/osascript -e 'do shell script "{os.path.join(os.path.dirname(sys.argv[0]),'Install.sh')} args 2>&1 etc" with administrator privileges'

here is Install.sh source code:

    #!/bin/sh
    /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)

I get this error output:

    Don't run this as root!

**So How can we install home-brew from a python script?**

If I run it as admin, it says "Don't run this as root!"
If I run it as non-admin, it says "Need sudo access on macOS"

**Comment by gromgit** [comment]
What problem are you trying to solve, that "run Homebrew install script from Python as root" is the (otherwise bizarre) answer?

**Comment by SMillerDev** [comment]
It needs sudo access but not all of it should be run as an admin. You need to somehow make it input sudo credentials while running, or (the better option) just reproduce what the script does in python if you need to setup homebrew with python.

**Reply by gromgit** [reply]
> You need to somehow make it input sudo credentials while running

Or "preload" sudo access via `sudo -v`, then every sudo invocation in `install.sh` shouldn't prompt for a password.

**Comment by reitermarkus** [comment]
You can set `SUDO_ASKPASS`.

**Comment by gauthierbuttez** [comment]
Thank you so much for trying to help me.


**If I run it as admin, it says "Don't run this as root!"**

**If I run it as non-admin, it says "Need sudo access on macOS"**

It is like French administration!!!

When you do A, they tell you "don't do A, do B"
When you do B, they tell you "don't do B, do A"

This is my Python script:


    def InstallHomeBrew(p_command):
        proc = subprocess.Popen(p_command, shell=True, stdout=True)
        (output, err) = proc.communicate()
        returncode_brew = proc.wait()
        print(f"returncode_brew : {returncode_brew}")
    
        if returncode_brew == 0:
            print("Homebrew installation was fine")
            print(f"Here is the succesfull command: {p_command}")
            return True
    
        else:
            print("Homebrew installation was not fine!")
            return False
    


    
    # === Get actual username ====================
    proc = subprocess.Popen("dscacheutil -q user | grep -A 3 -B 2 -e uid:\ 5'[0-9][0-9]'", shell=True,
                            stdout=subprocess.PIPE, stdin=subprocess.PIPE,
                            stderr=subprocess.STDOUT, close_fds=True)
    returncode_get_username = proc.wait()
    
    username = proc.stdout.readlines()
    # username = username.decode('utf-8')
    
    for line in username:
        if isinstance(line, bytes):
            line = line.decode()
    
        if line.find("name") != -1:
            mac_account = line.replace('name: ', '').replace('\n', '')


    # ====================================================================
    # Here below the list of different commands my function will try for installing homebrew
    
    command_install_homebrew_1 = f'/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"'
    command_install_homebrew_2 = f'sudo -u {mac_account} /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"'
    command_install_homebrew_3=f"""/usr/bin/osascript -e 'do shell script "{os.path.join(os.path.dirname(sys.argv[0]),'PhoneBot_Install.sh')} args 2>&1 etc" with administrator privileges'"""
    command_install_homebrew_4 = f'sudo -S /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"'
    command_install_homebrew_5=f"""/usr/bin/osascript -e 'do shell script "{os.path.join(os.path.dirname(sys.argv[0]),'PhoneBot_Install_as_Admin.sh')} args 2>&1 etc" with administrator privileges'"""
    command_install_homebrew_6 = f'sudo /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"'
    
    
    result_1 = InstallHomeBrew(command_install_homebrew_1)
    if not result_1:
        result_2 = InstallHomeBrew(command_install_homebrew_2)
        if not result_2:
            result_3 = InstallHomeBrew(command_install_homebrew_3)
            if not result_3:
                result_4 = InstallHomeBrew(command_install_homebrew_4)
                if not result_4:
                    result_5 = InstallHomeBrew(command_install_homebrew_5)
                    if not result_5:
                        result_6 = InstallHomeBrew(command_install_homebrew_6)
    



This is the source code of bash script:

    #Install.sh
    sudo -S /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
    
    #Install_as_Admin.sh
    sudo /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"


This is the output:

    Need sudo access on macOS (e.g. the user gauthierbtz to be an Administrator)!
    returncode_brew : 1
    Homebrew installation was not fine!
    Need sudo access on macOS (e.g. the user gauthierbtz to be an Administrator)!
    returncode_brew : 1
    Homebrew installation was not fine!
    Don't run this as root! (1)
    returncode_brew : 1
    Homebrew installation was not fine!
    Password:mypassword123
    Don't run this as root!
    returncode_brew : 1
    Homebrew installation was not fine!
    Don't run this as root! (1)
    returncode_brew : 1
    Homebrew installation was not fine!
    Don't run this as root!
    returncode_brew : 1
    Homebrew installation was not fine!
    
    Process finished with exit code 0


As you can see, they all failed!!!! Whatever I do!!!
**So How can we install home-brew from a python script?**

**Reply by gromgit** [reply]
And why does this script have to run as root? Or have to run `install.sh` as root?

**Reply by SMillerDev** [reply]
> If I run it as admin, it says "Don't run this as root!"

This is correct

> If I run it as non-admin, it says "Need sudo access on macOS"

Sudo access is not the same as running as root. Just like having access to a car is not the same as driving a car.

**Reply by gauthierbuttez** [reply]
Thanks @gromgit for trying to help.
But I don't understand you.

I clearly shared my code which shows I am executing the script as root and as non root with different methods.
I also published the output which show this crazy situation:

When you do A, they tell you "don't do A, do B"
When you do B, they tell you "don't do B, do A"

Could you please be a bit more accurate about your solution? Sorry you are not enough clear for me according to what I already tried and published.

**Reply by gauthierbuttez** [reply]
Thanks @SMillerDev for helping, so what is the solution?

**Reply by gromgit** [reply]
Only `command_install_homebrew_1` (the same command published on the Homebrew homepage) is needed. It's almost certainly failing because when `install.sh` invokes `sudo`, you'll get a password prompt, which isn't getting any input from your script.

That's why @SMillerDev said:
> You need to somehow make it input sudo credentials while running

to which I replied with one solution:
> Or "preload" sudo access via `sudo -v`

and @reitermarkus added a different one:
> You can set `SUDO_ASKPASS`.

My solution involves running `sudo -v` manually before running your script. `sudo` then caches the fact that you've authenticated recently, so all `sudo` calls in `install.sh` won't prompt for a password. That's great if you're typing stuff in at a terminal, but doesn't work so well for automated installs.

@reitermarkus 's solution involves setting the `SUDO_ASKPASS` environment variable to point to an askpass-style program that, say, prompts for a password via a GUI. The `sudo` calls in `install.sh` would then launch that program to authenticate the user for access.

Either way, please read the `sudo` man page for details. You'll need to understand how it works, to figure out the best solution for your circumstances.

**Reply by gauthierbuttez** [reply]
Thank you so much @gromgit for being more clear.

I did type the password in terminal when it prompted.
and I did try without sudo in my Install.sh

I can't use the sudo -v solution as my script will become a Mac OS .app which will be run by user. I can't ask user to type sudo -v to run my .app

the SUDO_ASKPASS solution, If I understood well, do the same than:
/usr/bin/osascript -e 'do shell script "{os.path.join(os.path.dirname(sys.argv[0]),'PhoneBot_Install.sh')} args 2>&1 etc" with administrator privileges'

it open a popup which ask the root password:

[https://prnt.sc/uc3vvw](https://prnt.sc/uc3vvw)

One thing I noticed, If I run the normal installation command manually from my terminal, it asked for password and then it start the installation process.

I don't understand why this can't succeed through a python script when the admin password is requested by popup!

**Reply by gromgit** [reply]
> I can't use the sudo -v solution as my script will become a Mac OS .app which will be run by user. I can't ask user to type sudo -v to run my .app

For future reference, these are the sort of details technical folks like myself are expecting when we ask "what problem are you trying to solve", because it's often the case that the solutions folks come up with don't actually solve the problems they have, or they haven't actually thought through their problems. Knowing these details can often help cut the time and effort involved in exploring useless pathways.

> the SUDO_ASKPASS solution, If I understood well, do the same than:
`/usr/bin/osascript -e 'do shell script "{os.path.join(os.path.dirname(sys.argv[0]),'PhoneBot_Install.sh')} args 2>&1 etc" with administrator privileges'`

No, they're very different.

`do shell script "install.sh" with administrator privileges` actually does the equivalent of `sudo install.sh`, which is very much _not_ supported by Homebrew.

`SUDO_ASKPASS` just specifies a program "hook" for the `sudo` calls in `install.sh` to get the necessary password _when it needs it_. (That said, I'm not familiar with askpass-style GUI programs for macOS, so can someone else add details?)

One other possibility: Launch a Terminal window, in which you run `install.sh`. Then the user can see the script's progress, and type in their password when they're prompted. This StackOverflow thread may help: https://stackoverflow.com/questions/989349/running-a-command-in-a-new-mac-os-x-terminal-window

**Reply by gromgit** [reply]
Continuing from where I left off...
> One thing I noticed, If I run the normal installation command manually from my terminal, it asked for password and then it start the installation process.

Yes, that's `sudo` being called by `install.sh`, specifically _and only_ to make stuff under `/usr/local` be writable by you as a normal user. The rest of the script is then run _as you_, so that `brew install` etc. after installation can be run _as you_.

> I don't understand why this can't succeed through a python script when the admin password is requested by popup!

Because the admin password popup is basically macOS (_not_ Homebrew) asking for permission to run _the entire `install.sh` as root_. `install.sh` blocks that because _all of Homebrew would then be installed as root_. Once that happens, _only root_ can run `brew install` etc.

Does that clarify everything we've written so far?
